> # CST1510 Programming for data communication and networks
> ## Week 2: Control Structures (Selection & Iteration) + Lists
> ## LAB ACTIVITY

<hr style="border:2px solid gray">


Welcome to your second Python lab session! This week, we'll focus on Python's **Conditional statements (if, elif, else)**, **loops (for, while)** and basic **list** operations.

By the end of this session, you'll be well-equipped to move forward in your Python journey.

### Topics Covered:

- **Learn to control the flow of your programs:** Understand how to make your code smarter by using conditional statements and loops.

- **Handle lists effectively:** Manage and manipulate data collections to create more efficient programs.


Once you’ve completed the tasks, select one or two activities demonstrating what you learned this week to include in your portfolio.

Remember, your portfolio should reflect your comprehension of variables, data types, and basic input/output operations—so choose wisely!
<hr style="border:1px solid gray">

# ACTIVITY:
<hr style="border:1px solid gray">

# Threshold:

1. **Password Strength Checker** 

**Scenario:**
You are a junior security analyst at a company. Employees often choose weak passwords, so you need to create a simple program to check password strength. Since you only know the basics of programming, your program will use simple rules.

**Requirements:**

1. Ask the user to enter a password.

2. If the password has fewer than 8 characters, print "Weak".

3. If the password has 8 or more characters, print "Moderate".

4. If the password has 12 or more characters, print "Strong"

In [1]:
# MISIS: M01052051
# Name: Jaden Gomes

password = input("Enter a password: ")

if len(password) < 8:
    print("Weak")
elif len(password) >= 8 and len(password) < 12:
    print("Moderate")
else:
    print("Strong")


Enter a password:  abc123


Weak


2. **Multiplication quiz**

Suppose you want to develop a program to help a first-grader practice multiplication. The program randomly generates two single-digit integers, number1 and number2, and displays to the student a question such as; What is 1 * 7. After the student types the answer, the program displays a message to indicate whether it is true or false.

To generate a random number, you can use the randint(a, b) function in the random module. This function returns a random integer i between a and b, inclusively. To obtain a random integer between 0 and 9, use randint(0, 9).

The program may be set up to work as follows:
Step 1: Randomly generate two single-digit integers for number1 (e.g., 4) and number 2 (e.g., 5)
Step 2: Prompt the student to answer, "What is 4 * 5?" (user input)
Step 3: Check whether the student’s answer is correct. 

In [2]:
# MISIS: M01052051
# Name: Jaden Gomes

import random

# Step 1: generate two random digits
num1 = random.randint(1, 9)
num2 = random.randint(1, 9)

# Step 2: ask the question
answer = int(input(f"What is {num1} * {num2}? "))

# Step 3: check answer
if answer == num1 * num2:
    print("Correct!")
else:
    print("Incorrect!")


What is 9 * 5?  45


Correct!


3. **Even or Odd** 
Write a program that reads an integer from the user.  Then your program should display a message indicating whether the integer is even or odd. 


In [3]:
# MISIS: M01052051
# Name: Jaden Gomes

num = int(input("Enter an integer: "))

if num % 2 == 0:
    print("Even")
else:
    print("Odd")


Enter an integer:  4


Even


4. **Body mass index (BMI)** is a measure of health and weight.

It can be calculated by taking your weight in kilograms and dividing it by the square of your height in meters. The interpretation of BMI for people 20 years or older is as follows:


BMI	Interpretation
BMI < 18.5	Underweight
18.5 ≤ BMI ≤ 25.0	Normal
25.0 ≤ BMI ≤ 30.0	Overweight
30.0 ≤ BMI 	Obese
 
Write a program that prompts the user to enter weight in pounds and height in inches and display the BMI. Note that one pound is 0.45359237 kilograms, and one inch is 0.0254 meters


In [4]:
# MISIS: M01052051
# Name: Jaden Gomes

weight_pounds = float(input("Enter weight in pounds: "))
height_inches = float(input("Enter height in inches: "))

weight_kg = weight_pounds * 0.45359237
height_m = height_inches * 0.0254

bmi = weight_kg / (height_m ** 2)

print(f"Your BMI is {bmi:.2f}")


Enter weight in pounds:  123.5
Enter height in inches:  65.7


Your BMI is 20.12


5. **Student Marks Analysis**
You are working as a data analyst in a school. You have the exam marks of students stored in a list. Your task is to analyze the data.

**Requirements**:

- Store marks of students in a list.

- Use a loop to calculate the average mark.

- Find the highest and lowest mark using list functions.

- Print the number of students who passed (marks ≥ 50) and failed.

In [5]:
# MISIS: M01052051
# Name: Jaden Gomes

marks = [55, 22, 80, 45, 69, 90, 30, 50]  # example list

average = sum(marks) / len(marks)
highest = max(marks)
lowest = min(marks)

passed = 0
failed = 0

for m in marks:
    if m >= 50:
        passed += 1
    else:
        failed += 1

print("Average mark:", average)
print("Highest mark:", highest)
print("Lowest mark:", lowest)
print("Passed:", passed)
print("Failed:", failed)


Average mark: 55.125
Highest mark: 90
Lowest mark: 22
Passed: 5
Failed: 3


6 . Display numbers from `-10` to `-1` using for loop

```Expected output:

-10
-9
-8
-7
-6
-5
-4
-3
-2
-1
```

In [6]:
# MISIS: M01052051
# Name: Jaden Gomes

for i in range(-10, 0):
    print(i)


-10
-9
-8
-7
-6
-5
-4
-3
-2
-1



7. Write a program to print the following number pattern using a loop



```Expected output:
1 
1 2 
1 2 3 
1 2 3 4 
1 2 3 4 5 
```


In [7]:
# MISIS: M01052051
# Name: Jaden Gomes

for i in range(1, 6):
    for j in range(1, i + 1):
        print(j, end=" ")
    print()


1 
1 2 
1 2 3 
1 2 3 4 
1 2 3 4 5 



8. Calculate the sum of all numbers from `1` to a `given number`

Write a program to accept a number `n` from a user and calculate the sum of all numbers from 1 to a given number n


For example, if the user entered 10 the output should be 55 (1+2+3+4+5+6+7+8+9+10)

```
Expected output
Enter the value n: 10
The sum of all numbers until 10 is 55
```




In [8]:
# MISIS: M01052051
# Name: Jaden Gomes

n = int(input("Enter the value of n: "))

total = 0
for i in range(1, n + 1):
    total += i

print("The sum of all numbers until", n, "is", total)


Enter the value of n:  5


The sum of all numbers until 5 is 15


9. Write a program to print multiplication table of a given number

For example, `num = 2`

```
Expected output:

0 x 2 = 0
1 x 2 = 2
2 x 2 = 4
3 x 2 = 6
4 x 2 = 8
5 x 2 = 10
6 x 2 = 12
7 x 2 = 14
8 x 2 = 16
9 x 2 = 18
```



In [9]:
# MISIS: M01052051
# Name: Jaden Gomes

num = int(input("Enter a number: "))

for i in range(0, 10):
    print(f"{i} x {num} = {i * num}")


Enter a number:  5


0 x 5 = 0
1 x 5 = 5
2 x 5 = 10
3 x 5 = 15
4 x 5 = 20
5 x 5 = 25
6 x 5 = 30
7 x 5 = 35
8 x 5 = 40
9 x 5 = 45


10. Displays the following multiplication table. Consider the spacing between digits

```
Expected output:

0   0   0   0   0   0   0   0   0   0 
0   1   2   3   4   5   6   7   8   9 
0   2   4   6   8  10  12  14  16  18 
0   3   6   9  12  15  18  21  24  27 
0   4   8  12  16  20  24  28  32  36 
0   5  10  15  20  25  30  35  40  45 
0   6  12  18  24  30  36  42  48  54 
0   7  14  21  28  35  42  49  56  63 
0   8  16  24  32  40  48  56  64  72 
0   9  18  27  36  45  54  63  72  81 
```



In [10]:
# MISIS: M01052051
# Name: Jaden Gomes

for i in range(1, 10):
    for j in range(1, 10):
        print(f"{i*j:3}", end=" ")
    print()


  1   2   3   4   5   6   7   8   9 
  2   4   6   8  10  12  14  16  18 
  3   6   9  12  15  18  21  24  27 
  4   8  12  16  20  24  28  32  36 
  5  10  15  20  25  30  35  40  45 
  6  12  18  24  30  36  42  48  54 
  7  14  21  28  35  42  49  56  63 
  8  16  24  32  40  48  56  64  72 
  9  18  27  36  45  54  63  72  81 


11. Use the math.sqrt function. Write a program that prints the following table using the sqrt function in the math module.

```
Number      Square root
0   	 	0.00
2    		1.41
4    		2.00
6    		2.45
8    		2.83
10    		3.16
12    		3.46
14    		3.74
16    		4.00
18    		4.24
```


In [11]:
# MISIS: M01052051
# Name: Jaden Gomes

import math

print("Number\tSquare root")
for n in range(0, 19, 2):
    print(f"{n}\t{math.sqrt(n):.2f}")


Number	Square root
0	0.00
2	1.41
4	2.00
6	2.45
8	2.83
10	3.16
12	3.46
14	3.74
16	4.00
18	4.24


# Typical: 

12. **Dog Years**

It is commonly said that one human year is equivalent to 7 dog years. However, this simple conversion fails to recognise that dogs reach adulthood in approximately two years. As result, some people believe that it is better to count each of the first two human years as 10.5 dog years, and then count each additional human year as 4 dog years. 

Write a program that implements the conversion from human years to dog years described in the previous paragraph. Ensure that your program works correctly for conversions of less than two or more humans’ years. Your program should display an appropriate error message if the user enters a negative number. 



13. **Vowel or Constant**

In this exercise you will create a program that read a letter of the alphabet from the user. If the user enter a, e, i, o or u then your program should display a message indicating that the entered letter is a vowel. If the user enters y then your program should display a message indicating that sometimes y is a vowel, and sometimes y is a consonants.  Otherwise your program should display a message indicating that the letter is a constant. 


14. Print the following pattern, write a program to use for loop to print the following reverse number pattern

```
Expected output:
5 4 3 2 1 
4 3 2 1 
3 2 1 
2 1 
1
```


15. **Find the highest score**

Write a program that prompts the user to enter the number of students and each student’s name and score and display the student name and highest score.


16. Write a **guessing game** in Python where the user has to guess a secret number. After every guess the program tells the user whether their number was too large or too small. At the end the number of tries needed should be printed.


17. **Collecting Positive Numbers**
You are writing a small program to help a teacher keep track of positive numbers entered by students. The teacher also wants to know if any even numbers were given.

**Requirements:**

1. Keep asking the user to enter a positive number, or type `"done"` to stop.
2. Store all entered numbers in a **list**.
3. After the input ends:

   * If numbers were entered:

     * Print all the numbers entered.
     * Check the list and print the **first even number**. If no even number exists, print `"No even numbers found"`.
   * If no numbers were entered, print `"No numbers were entered"`.


> ### Note 
>
>* **`lower()`**: Converts any text into lowercase.
>  Example: `"DONE".lower()` becomes `"done"`.
>  This makes it easier to compare user input without worrying about uppercase/lowercase.
>
>* **`isdigit()`**: Checks if a string only contains digits (0–9).
>  Example: `"123".isdigit()` returns `True`, but `"12a".isdigit()` returns `False`.
>  This helps us make sure the user typed a valid number before converting it into an integer.
> 


---

# Excellent:

18. Suppose you want to develop a **program to play a lottery.**
    
The program randomly generates a two-digit number, prompts the user to enter a <u>two-digit number</u>, and determines whether the user wins according to the following rules:


    1. If the user’s input matches the lottery in the exact order, the award is £10,000.
    2. If all the digits in the user’s input match all the digits in the lottery number, the award is £3,000.
    3. If one digit in the user’s input matches a digit in the lottery number, the award is £1,000.




19. The program prompts the user to enter a <u>three-digit number</u> and determines whether the user wins according to the following rules:

    1. If the user input matches the lottery number in the exact order, the award is £10,000.
    2. If all the digits in the user input match all the digits in the lottery number, the award is £3,000.
    3. If one digit in the user input matches a digit in the lottery number, the award is £1,000.

20. (Find the two highest scores) - Write a program that prompts the user to enter the number of students and each student’s score and displays the highest and second highest scores.



21. **Password cracker** - write a program to print the accuracy of password guess.

For example: The actual password is `MDX123` while a user tried `MDX321` or `mdx123`, the program should tell that how accurately the password was guessed, in this case it should be `50%` accuracy. if a user enter `MDX124`, the output should be `83%` and in case of `123MDX` the output should be `0%`.

Points to keep in mind while calculating the accuracy a) case of the letters entered b) order of the letters

If a password is accurately guessed, the output should state `Password cracked!`

---

## Portfolio
Select one or two completed activities that best demonstrate this week’s learning. Polish the code (naming, formatting, comments) so you can commit them to Git in later weeks
